core-processor/src/main/java/io/micronaut/inject/writer/AbstractClassFileWriter.java

```diff
@@ -237,19 +237,28 @@ protected static void pushTypeArgumentElements(
             return;
         }
         Set<String> visitedTypes = new HashSet<>(5);
-        pushTypeArgumentElements(owningType, owningTypeWriter, generatorAdapter, declaringElementName, types, visitedTypes, defaults, loadTypeMethods);
+        pushTypeArgumentElements(owningType, owningTypeWriter, generatorAdapter, null, declaringElementName, types, visitedTypes, defaults, loadTypeMethods, false);
     }
 
     private static void pushTypeArgumentElements(
             Type owningType,
             ClassWriter declaringClassWriter,
             GeneratorAdapter generatorAdapter,
+            @Nullable TypedElement declaringElement,
             String declaringElementName,
             Map<String, ClassElement> types,
             Set<String> visitedTypes,
             Map<String, Integer> defaults,
-            Map<String, GeneratorAdapter> loadTypeMethods) {
-        if (visitedTypes.contains(declaringElementName)) {
```

Comment:
In case of `List<List>`, type parameter for type `List` is already contained, therefore have to change the check.

Instead I use `classElement.isGenericPlaceholder() || classElement.isTypeVariable()` which corresponds to cases when the code uses generic type variable (such as `List<T>` including bound `T extends MyClass`).